paprika/tleap.py

```diff
@@ -0,0 +1,574 @@
+import os as os
+import re as re
+import subprocess as sp
+
+import logging as log
+import numpy as np
+import parmed as pmd
+from parmed.structure import Structure as ParmedStructureClass
+from paprika import utils
+
+
+N_A = 6.0221409 * 10 ** 23
+ANGSTROM_CUBED_TO_LITERS = 1 * 10 ** -27
+
+class System(object):
+    """
+    Class for building AMBER prmtop/inpcrd files with tleap.
+
+    Class Variables
+    ---------------
+    template_file : str
+        Name of template file to read boilerplate `tleap` input (e.g., `source leaprc`... lines). Default: None
+    template_lines : lst
+        The list of tleap commands which are needed to build the system. Either the template_file or
+        the template_lines needs to be set prior to running build(). Default: None
+    loadpdb_file : str
+        If specified, this PDB file will be loaded via loadpdb instead of whatever file is specified
+        in the template. Default: None
+    pbc_type : str
+        Type of solvation (cubic, rectangular, octahedral, or None). Default: cubic
+    buffer_target : int or str
+        The desired amount of water to add. Either an integer number of waters, or a buffer
+        size ending in with an 'A', e.g. '12.0A'
+    water_box : str
+        The type of water box, e.g., TIP3PBOX (see AMBER manual for acceptable values). Default: TIP3PBOX
+    neutralize : bool
+        Whether to neutralize the system with counterions. Default: True
+    counter_cation : str
+        If neutralize=True, this positive ion will be used. Default: Na+
+    counter_anion : str
+        If neuralize=True, this negative ion will be used. Default: Cl-
+    add_ions : list
+        A list of additional ions to be added to the system, specified by the residue name and 
+        an indication of the amount: an integer number, the molarity (M), or the molality (m).
+        For example, the following list shows valid examples:
+        add_ions = ['MG', 2, 'Cl-', 4, 'K', '0.150M', 'BR', '0.150M', 'NA', '0.050m', 'F', '0.050m']
+        Default: None
+    output_path : str
+        Directory path where tleap input files will be written and executed. Associated input
+        files (frcmod, mol2, pdb, etc) need to present in that path.
+    output_prefix : str
+        Append a prefix to files created by this module. Default: 'build'
+
+    Advanced Class Variables
+    ------------------------
+    unit : str
+        The tleap unit name. Default: 'model'
+    buffer_value : float
+        The initial buffer_value to be attempted for solvation. Default: 1.0
+    target_waters : int
```

Comment:
Maybe "advanced" isn't the right word .... "low level" might be better?  `buffer_target` is what we give to users to set as either a integer number of waters or a buffer values "12.0A".  Then `target_waters` gets set.

Reply:
Just seems to me like the two things users will care most are `buffer_value` and `target_waters`, but it's not worth litigating the semantics.

Reply:
Are you trying to say the current set up is confusing?  Right now users set buffer_target,  which is either an integer, or a float with a letter A after. That  value gets processed and sets target_waters and buffer_value.  We could switch so that there is no such thing as a buffer_target, and have the users directly set either offer value or target_waters.

Reply:
Yes, I'm not sure the use-case for `buffer_target`.

Reply:
Working on removing it .....